### PR TITLE
Implement gpu_scheduling_behavior parameter and basic functionality

### PR DESCRIPTION
**Pull request by meln1k**
Summary: Implemented a new GPU-aware offer matching logic, command line flags, RAML examples and unit test

JIRA issues: MARATHON-8088

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/MarathonConf.scala [hunk 1]
```diff
@@ -317,5 +317,26 @@ trait MarathonConf
...
+      "gpu_resources must be set in order to use gpu_scheduling_behavior"
+    )
+    require(
+      allowedSettings.contains(setting),
+      s"Setting $setting is invalid. Valid settings are ${allowedSettings.mkString(", ")}"
+    )
+    true
+  }
+
+  lazy val gpuSchedulingBehavior = opt[String](
```
Comment (alenkacz) [inline src/main/scala/mesosphere/marathon/MarathonConf.scala]: should we maybe do also `noshort = true` here?

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 2]
```diff
@@ -236,13 +248,40 @@ object ResourceMatcher extends StrictLogging {
...
+          }
+          true
+
+        case GpuSchedulingBehavior.Restricted =>
+          val noPersistentVolumeToMatch = PersistentVolumeMatcher.matchVolumes(offer, reservedInstances).isEmpty
+          if (gpuResourcesAreWasted && noPersistentVolumeToMatch) {
+            noOfferMatchReasons += NoOfferMatchReason.DeclinedScarceResources
+            false
+          } else {
+            true
```
Comment (alenkacz) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: should we maybe log something here? That we are launching on GPU node while restricted but we are doing that because of persistent volume?

File: src/test/scala/mesosphere/marathon/raml/PodStatusConversionTest.scala [hunk 3]
```diff
@@ -407,6 +407,30 @@ class PodStatusConversionTest extends UnitTest {
           "persistent#volume#5425cbaa-8fd3-45f0-afa4-74ef4fcc594b")))
     }
 
+    "persistent pod launched, received RUNNING status from Mesos, task endpoint health looks great" in {
```
Comment (alenkacz) [inline src/test/scala/mesosphere/marathon/raml/PodStatusConversionTest.scala]: how is this test related to GPUs?

File: src/main/scala/mesosphere/mesos/NoOfferMatchReason.scala [hunk 4]
```diff
@@ -14,6 +14,7 @@ object NoOfferMatchReason {
   case object UnfulfilledConstraint extends NoOfferMatchReason
   case object AgentMaintenance extends NoOfferMatchReason
   case object NoCorrespondingReservationFound extends NoOfferMatchReason
+  case object DeclinedScarceResources extends NoOfferMatchReason
```
Comment (alenkacz) [inline src/main/scala/mesosphere/mesos/NoOfferMatchReason.scala]: you have to add this to the reasonfunnel otherwise it won't appear in the API. I've been there and it sucks and is easy to miss...

Maybe it would be nice to add a test that verifies that it appears in API, then we would see this...

File: src/test/scala/mesosphere/marathon/raml/PodStatusConversionTest.scala (hunk 3 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/raml/PodStatusConversionTest.scala]: It's something from another changelist :) good catch!

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala (hunk 2 above)
Comment (meln1k) [reply src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: Design doc says nothing about it. @timcharper what do you think?

Comment (alenkacz) [reply src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: I did not mean warn or error, I would log just info... But @timcharper feel free to comment :)

File: src/main/scala/mesosphere/mesos/ResourceMatcher.scala [hunk 5]
```diff
@@ -236,13 +248,42 @@ object ResourceMatcher extends StrictLogging {
...
+          }
+          true
+
+        case GpuSchedulingBehavior.Restricted =>
+          val noPersistentVolumeToMatch = PersistentVolumeMatcher.matchVolumes(offer, reservedInstances).isEmpty
+          if (gpuResourcesAreWasted && noPersistentVolumeToMatch) {
+            noOfferMatchReasons += NoOfferMatchReason.DeclinedScarceResources
+            false
+          } else {
+            addOnMatch(() => logger.warn(s"Runspec [${runSpec.id}] doesn't require any GPU resources but " +
```
Comment (alenkacz) [inline src/main/scala/mesosphere/mesos/ResourceMatcher.scala]: I would prefer to log this as info

File: docs/docs/rest-api/public/api/v2/examples/queue.json [hunk 6]
```diff
@@ -145,9 +145,9 @@
             "processed": 23
           },
           {
-            "reason": "InsufficientGpus",
+            "reason": "DeclinedScarceResources",
```
Comment (ichernetsky) [inline docs/docs/rest-api/public/api/v2/examples/queue.json]: Should the examples with `InsufficientGpus` be kept instead of being deleted?

Comment (alenkacz) [reply docs/docs/rest-api/public/api/v2/examples/queue.json]: they were in the example two times before this change ... I actually wondered the same when reviewing :)

Comment (meln1k) [reply docs/docs/rest-api/public/api/v2/examples/queue.json]: yep, it seems it was a copy paste error before
